### ISSUE TITLE
Dial error on windows 7

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ipfs#4512]
#### Version information:

0.4.14-dev
#### Type:
Bug
#### Severity:
Critical
#### Description:

No peers in swarm. When dialing, @djdv gets:

```
Error: connect QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3 failure: dial attempt failed: <peer.ID WABRCq> --> <peer.ID SoLMeW> dial attempt failed: dial tcp4 0.0.0.0:4737->178.62.61.185:4001: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
```

Caused by something in #4453.

**Comment by Stebalien** [comment]
@djdv could you try disabling reuseport (set the environment variable `IPFS_REUSEPORT=false`). It's not a fix but it'll help me determine what's going on here.

**Comment by gdeverlant** [comment]
I come up with another error with my tests on Windows 7. When I check the HTTP API via Chrome the requests are returning results.

When I try via my video editor to access the HTTA API via the URLLoader class the IPFS node returns 403 - Forbidden error.

![image](https://user-images.githubusercontent.com/26530975/34258813-f989cdf2-e65f-11e7-981d-2c71acc442a5.png)

**Comment by gdeverlant** [comment]
Tried with the HTTPService API

![image](https://user-images.githubusercontent.com/26530975/34260754-2d01ec76-e667-11e7-9ca2-a7d4d3c8d893.png)

**Comment by gdeverlant** [comment]
Accessing the API is working via Socket TCP Programming only...

![image](https://user-images.githubusercontent.com/26530975/34269103-6e4f2948-e683-11e7-9336-b3f5610c2b4a.png)

**Comment by Stebalien** [comment]
@gdeverlant your errors appear to be unrelated to this issue. Please file a new one (or open a thread on https://discuss.ipfs.io as this doesn't look like a bug).

**Comment by djdv** [comment]
@Stebalien 
For context I'm on Windows 10, this may be affecting all versions of Windows, since I'm seeing other users having empty peer tables as well on different versions, when using IPFS past #4453.

Setting IPFS_REUSEPORT to false doesn't seem to have any effect for me, I set it, start the daemon (built from current master), and try to connect to a peer, but encounter the same error as before. I just ran this with Go 1.10 but I was getting the same results on 1.9.2 a few days ago.

Promt 1
```
>ipfs version --all
go-ipfs version: 0.4.14-dev-ce22b83f2
Repo version: 6
System version: amd64/windows
Golang version: go1.10beta1

>SET IPFS_REUSEPORT=false

>ipfs daemon
...
```
Promt 2
```
>ipfs swarm peers

>ipfs swarm connect /ip4/178.62.61.185/tcp/4001/ipfs/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3
Error: connect QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3 failure: dial attempt failed: <peer.ID WABRCq> --> <peer.ID SoLMeW> dial attempt failed: dial tcp4 0.0.0.0:4737->178.62.61.185:4001: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
```

**Comment by Stebalien** [comment]
Got it. Is there any way on windows to see if there is an existing connection on that port/IP? Any way to see if a connection is established (wireshark?).

@whyrusleeping can you reproduce in your VM?

**Comment by djdv** [comment]
`netstat -aon | findstr %port%` will show ports in use and their state, I find [process hacker](http://processhacker.sourceforge.net) to be useful in monitoring as well. I run ipfs on the old default port 4737. Checking with port reuse enabled, 4001 shows almost everything in a TIME_WAIT state except for the interfaces I am listening on (0.0.0.0:4737, [::]:4737) which are correctly in the LISTENING state, 4737 shows a single connection in the TIME_WAIT state (local 4737 waiting on a remote 4001)
Disabling reuse gives me nothing for port 4001 and again the 2 interfaces for 4737, nothing else.

I'm pretty confident that nothing on this machine should be using either of these ports, nothing is configured to and nothing is showing for it, switching to 4001 didn't resolve it for me either.

**Comment by Stebalien** [comment]
I'm getting this on linux as well. It can be readily reproduced with the go-libp2p echo example. I'm also seeing the `TIME_WAIT` stuff so I think a connection is opening, then closing, but not fully closing.

**Comment by gdeverlant** [comment]
Can someone fix this problem ASAP because we can't build and integrate our product to IPFS.

**Comment by djdv** [comment]
@gdeverlant are you encountering this issue on builds before #4453?
If you need ipfs on Windows right now, you might want to try using the release build v0.4.13 or using a source base without dfe7ef4fcfaf340846ee16031550601b3ff625d6 until this is resolved.

**Comment by gdeverlant** [comment]
it's the first time I'm using ipfs. I didn't used any other versions yet!

**Comment by gdeverlant** [comment]
I just did tried your suggestion and here is the problems:

It is better than before but still there are lots of error messages.
The peers are listed, the webui works well but when I want to see the peers the UI hangs and chrome sucks 50 % of my cpu.

**Comment by Stebalien** [comment]
Unfortunately, that's an unrelated webui bug. The webui's peer display isn't optimised for thousands of peers.

**Comment by Stebalien** [comment]
> It is better than before but still there are lots of error messages.

Please post them. Unfortunately, we have a lot of noisy error messages (that you can mostly ignore).

**Comment by Stebalien** [comment]
> I'm getting this on linux as well. It can be readily reproduced with the go-libp2p echo example. I'm also seeing the TIME_WAIT stuff so I think a connection is opening, then closing, but not fully closing.

Not the issue.

I believe the issue is that non-reuseport dialing has been broken by 5ebce0ea21d3e12812e133e9ab2127376940ec7a. We now try to set the source port when dialing, even when not dialing with reuseport, which obviously fails if we're already using that port (e.g., listening on it). Fix incoming.

**Comment by Stebalien** [comment]
Fix coming in https://github.com/libp2p/go-tcp-transport/pull/20

**Comment by Stebalien** [comment]
Could one of you test [/ipfs/Qma9vnYd4LV7DJkcsx4on3ezzuHT5JQcU2J2GaS3jmeKCt](https://ipfs.io/ipfs/Qma9vnYd4LV7DJkcsx4on3ezzuHT5JQcU2J2GaS3jmeKCt)? It should fix this.

**Comment by djdv** [comment]
@Stebalien 
It appears to make connections (my peer table fills up) and the gateway can pull data from me.

**Comment by gdeverlant** [comment]
Is this resolved ? Can I try this and implement IPFS for my invention ?

**Comment by Stebalien** [comment]
> It loads with only 2 errors:

Those errors are "normal" and should go away in the near future.

> But the hell load of error happens when I try to go in the webui and check the peers.

That's because the webui has some *really* unoptimized code rendering all connected peers on a globe. For now, the recommended way to check which peers you're connected to is to run `ipfs swarm peers` on the command line.

> I checked the G:\ipfs folder's properties and it seems to be set to read-only

That would explain that list of errors but I'm pretty sure we don't set the repo to read-only (unfortunately, I don't use windows so it's a bit hard for me to check). Did it start out read-only (did it work and then stop working) or did it suddenly become read-only?

**Comment by djdv** [comment]
@gdeverlant 
The read only box there is set to the tri-state, it is most likely representing "unset" or "mixed", a read-only state looks like a checkmark ✔️ in that box.

There may be something else preventing you from writing to the datastore, without information on how ipfs was invoked and the permissions for the G: volume, it's hard to say.

I believe the dht cast errors are already known, I think they are being demoted to warnings not errors. Someone correct me if this is wrong, please.

**Comment by Stebalien** [comment]
I've moved this new discussion to #4528 (I like to keep issues clean).

**Comment by gdeverlant** [comment]
@Stebalien If you don't use Windows just go get a Virtualbox image and start playing with.

**Comment by djdv** [comment]
@Stebalien 
I've been running the test binary from above (Qma9vnYd4LV7DJkcsx4on3ezzuHT5JQcU2J2GaS3jmeKCt) for a while now, this morning I returned to a panic.
```
panic: runtime error: slice bounds out of range

goroutine 1414 [running]:
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table.(*blockIterator).parseKV(0xc12327ac60, 0x11802ab3ef, 0xc000710003)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table/iterator.go:118 +0x54d
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table.(*blockIterator).Next(0xc12327ac60)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table/iterator.go:154 +0x1ec
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table.(*Iterator).next(0xc0cd84c360)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table/iterator.go:347 +0xc0
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table.(*Iterator).Next(0xc0cd84c360)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table/iterator.go:397 +0x4d
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table.(*ConcatIterator).Next(0xc0d2310a50)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/table/iterator.go:507 +0x3a
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/y.(*MergeIterator).Next(0xc0d2310be0)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/y/iterator.go:216 +0x5b
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger.(*Iterator).parseItem(0xc0cf33a4d0, 0x1)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/iterator.go:388 +0x71c
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger.(*Iterator).prefetch(0xc0cf33a4d0)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/iterator.go:460 +0x9b
gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger.(*Iterator).Seek(0xc0cf33a4d0, 0xc0b0f54c80, 0x42, 0x50)
        /home/steb/projects/go/src/gx/ipfs/QmXtSA4dxoDKMzhHkwqeoBoBYASov1ZWnLQo38juKNbe3z/badger/iterator.go:492 +0x3ee
gx/ipfs/Qmdin8YL17fL1BC5ej6o9b8es6MBoiQjKVdyxEwJh3HVmf/go-ds-badger.(*datastore).QueryNew(0xc042039350, 0xc0b0f54be0, 0x42, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/steb/projects/go/src/gx/ipfs/Qmdin8YL17fL1BC5ej6o9b8es6MBoiQjKVdyxEwJh3HVmf/go-ds-badger/datastore.go:144 +0x1ef
gx/ipfs/Qmdin8YL17fL1BC5ej6o9b8es6MBoiQjKVdyxEwJh3HVmf/go-ds-badger.(*datastore).Query(0xc042039350, 0xc0b0f54be0, 0x42, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/steb/projects/go/src/gx/ipfs/Qmdin8YL17fL1BC5ej6o9b8es6MBoiQjKVdyxEwJh3HVmf/go-ds-badger/datastore.go:133 +0x74
gx/ipfs/QmU7tt6mHJ5Wocjy2omBxpDfN8g9pkRimzJae7EXdrs96k/go-ds-measure.(*measure).Query(0xc04213c5a0, 0xc0b0f54be0, 0x42, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/steb/projects/go/src/gx/ipfs/QmU7tt6mHJ5Wocjy2omBxpDfN8g9pkRimzJae7EXdrs96k/go-ds-measure/measure.go:141 +0x12e
gx/ipfs/QmU7tt6mHJ5Wocjy2omBxpDfN8g9pkRimzJae7EXdrs96k/go-ds-measure.(*measure).Query(0xc04213c6c0, 0xc0b0f54be0, 0x42, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/steb/projects/go/src/gx/ipfs/QmU7tt6mHJ5Wocjy2omBxpDfN8g9pkRimzJae7EXdrs96k/go-ds-measure/measure.go:141 +0x12e
gx/ipfs/QmXVJax39Hpq2KHYsJZobhVSyKNNwWp3pXLjmdoqYheDHA/autobatch.(*Datastore).Query(0xc0b3e48140, 0xc0b0f54be0, 0x42, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/steb/projects/go/src/gx/ipfs/QmXVJax39Hpq2KHYsJZobhVSyKNNwWp3pXLjmdoqYheDHA/autobatch/autobatch.go:80 +0xd4
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers.loadProvSet(0x1490720, 0xc0b3e48140, 0xc0c6aed770, 0x0, 0x0, 0xc0c7527f00)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers/providers.go:115 +0xd9
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers.(*ProviderManager).addProv(0xc0b3334180, 0xc0c6aed770, 0xc0d07daae0, 0x22, 0x0, 0x0)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers/providers.go:167 +0x361
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers.(*ProviderManager).run(0xc0b3334180)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers/providers.go:256 +0xae5
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers.NewProviderManager.func1(0x1494500, 0xc0b3334240)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/providers/providers.go:73 +0x31
gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go.func1(0xc090f08040, 0xc0b3334240, 0xc0b33342a0)
        /home/steb/projects/go/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:112 +0x43
created by gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go
        /home/steb/projects/go/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:111 +0x2d8
```

**Comment by whyrusleeping** [comment]
@djdv well thats another issue entirely. looks like a potential badger bug

**Comment by whyrusleeping** [comment]
@djdv do you still have the repo from this failure? If so, how big is it and  would you be willing to share it with us? The badger team wants to look at the badger files to try and debug the panic you posted.

**Comment by djdv** [comment]
@whyrusleeping I still have it but it in an active repo, which has since been modified via browsing other users hashes and adding things myself. I'm willing to give some form of remote access to it or upload a copy somewhere else (whichever is easier) but it's relatively large (178GB), as such I will contact you directly via email about this when I get a chance later.

**Comment by djdv** [comment]
2nd panic with above build
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x400 pc=0x40c724]

goroutine 39750634 [running]:
gx/ipfs/QmYijbtjCxFEjSXaudaQAUz3LN5VKLssm8WCUsRoqzXmQR/go-libp2p-peerstore.(*AddrManager).AddAddrs(0xc04253c890, 0xc0cf39e000, 0x22, 0xc0be4a1c00, 0xa9, 0xa9, 0x2540be400)
        /home/steb/projects/go/src/gx/ipfs/QmYijbtjCxFEjSXaudaQAUz3LN5VKLssm8WCUsRoqzXmQR/go-libp2p-peerstore/addr_manager.go:124 +0x2e2

gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht.(*IpfsDHT).closerPeersSingle(0xc04215a300, 0x148ddc0, 0xc0d5bc30b0, 0xc0bd97fb90, 0x26, 0xc0bd021830, 0x22, 0x10, 0xd85800, 0xc0bd021801, ...)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/lookup.go:117 +0x1cc
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht.(*IpfsDHT).GetClosestPeers.func1(0x148ddc0, 0xc0d5bc30b0, 0xc0bd021830, 0x22, 0x1, 0x1, 0xc0f2450160)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/lookup.go:65 +0x128
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht.(*dhtQueryRunner).queryPeer(0xc0c2ffc780, 0x1494500, 0xc0eb19bb00, 0xc0bd021830, 0x22)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/query.go:267 +0x140
gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht.(*dhtQueryRunner).spawnWorkers.func1(0x1494500, 0xc0eb19bb00)
        /home/steb/projects/go/src/gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht/query.go:208 +0x57
gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go.func1(0xc0d7c5ea20, 0xc0eb19bb00, 0xc0eb19bc20)
        /home/steb/projects/go/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:112 +0x43
created by gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess.(*process).Go
        /home/steb/projects/go/src/gx/ipfs/QmSF8fPo3jgVBAy8fpdjjYqgG87dkJgUprRBHRd2tmfgpP/goprocess/impl-mutex.go:111 +0x2d8
```

**Comment by Stebalien** [comment]
That's a (bad) memory corruption. Something is shitting all over memory and I'm not convinced it's badger. Possibly related to #4483, #4527.

**Comment by Stebalien** [comment]
Fixed.
